feat: drop `cachetools` dependency in favor of simple local implementation

akx opened the pull request:
This PR drops the dependency on `cachetools`, which was only used for its LRU cache class, in favor of a simple local implementation.

This should have a small but positive effect on many downstream users given how many times this library is downloaded per day.

akx committed:
feat: drop `cachetools` dependency in favor of simple local implementation

akx commented:
@clundin25 Unfortunately not. The transparency of a `lru_cache`d function is not really compatible with how the client code here is e. g. doing `if foo in cache:`.

westarle commented:
How big is the change to update the client code so it can use the standard library and avoid maintaining this implementation?

/cc @sai-sunder-s

akx commented:
> How big is the change to update the client code so it can use the standard library and avoid maintaining this implementation?

Sorry, I wasn't sure whether that was directed at me or the person cc'd.

Not trivial, since it depends on being able to introspect the cache entry before deciding whether it's usable. `lru_cache` does not expose such a mechanism (nor being able to replace a single key from the cache, which is also required).

```python
token, expiry = self._cache.get(audience, (None, None))
if token is None or expiry < _helpers.utcnow():
```

akx commented:
Rebased again. Could this be moved along?

akx commented:
Rebased (post #1773 merge).

SchubmannM commented:
Would be great to get this is as indeed it's blocking other projects from updating `cachetools` beyond 6.0.0

chalmerlowe commented on google/auth/_cache.py:
```diff
@@ -0,0 +1,42 @@
+from collections import OrderedDict
```
```suggestion
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from collections import OrderedDict
```

chalmerlowe commented on google/auth/_cache.py:
```diff
@@ -0,0 +1,42 @@
+    def clear(self):
+        super().clear()
+        self._order.clear()
+
```
```suggestion
    def get(self, key, default=None):
        try:
            value = super().__getitem__(key)
            self._update(key)
            return value
        except KeyError:
            return default
```

chalmerlowe commented on tests/test__cache.py:
```diff
@@ -0,0 +1,22 @@
+def test_zero_size_lru_cache():
+    lru_cache = LRUCache(0)
+    lru_cache["a"] = 1
+    assert "a" not in lru_cache
```
```suggestion
def test_lru_cache():
    """Test the LRUCache for generally expected functionality and ordering."""
    lru_cache = LRUCache(2)
    lru_cache["a"] = 1
    lru_cache["b"] = 2
    assert lru_cache["a"] == 1
    lru_cache["c"] = 3
    assert "b" not in lru_cache
    assert lru_cache["a"] == 1
    assert lru_cache["c"] == 3
    lru_cache["d"] = 4
    assert "a" not in lru_cache
    assert lru_cache["c"] == 3
    assert lru_cache["d"] == 4


def test_zero_size_lru_cache():
    """Confirm the LRUCache handles zero-size correctly."""
    lru_cache = LRUCache(0)
    lru_cache["a"] = 1
    assert "a" not in lru_cache


def test_lru_cache_get_updates_lru():
    """Confirm the LRUCache handles get calls correctly."""
    lru_cache = LRUCache(2)
    lru_cache["a"] = 1
    lru_cache["b"] = 2
    
    # Access "a" via get(), making it MRU.
    assert lru_cache.get("a") == 1
    
    # Add "c", which should evict "b" (LRU), not "a".
    lru_cache["c"] = 3
    
    assert "a" in lru_cache
    assert "b" not in lru_cache
    assert "c" in lru_cache


def test_lru_cache_get_missing():
    """Confirm the LRUCache handles missing keys correctly."""
    lru_cache = LRUCache(2)
    assert lru_cache.get("missing") is None
    assert lru_cache.get("missing", "default") == "default"
```

chalmerlowe committed:
fix: add `.get()` to LRUCache

Co-authored-by: Aarni Koskela <akx@iki.fi>

chalmerlowe commented on tests/test__cache.py:
```diff
@@ -0,0 +1,48 @@
+    """Confirm the LRUCache handles missing keys correctly."""
+    lru_cache = LRUCache(2)
+    assert lru_cache.get("missing") is None
+    assert lru_cache.get("missing", "default") == "default"
```
```suggestion
    assert lru_cache.get("missing", "default") == "default"
    
    
def test_lru_cache_clear():
    """Confirm the LRUCache clears the cache properly."""
    lru_cache = LRUCache(2)
    lru_cache["a"] = 1
    lru_cache["b"] = 2
    assert len(lru_cache) == 2
    
    lru_cache.clear()
    assert len(lru_cache) == 0
    assert "a" not in lru_cache
    assert "b" not in lru_cache
    # Ensure internal order is also cleared
    assert len(lru_cache._order) == 0


def test_lru_cache_delitem():
    """Confirm the LRUCache deletes individual items properly."""
    lru_cache = LRUCache(2)
    lru_cache["a"] = 1
    lru_cache["b"] = 2
    
    del lru_cache["a"]
    assert "a" not in lru_cache
    assert len(lru_cache) == 1
    # Ensure it's removed from internal order
    assert "a" not in lru_cache._order
    
    # Test that we can continue using the cache
    lru_cache["c"] = 3
    assert "c" in lru_cache
    assert "b" in lru_cache
    assert len(lru_cache) == 2
```

chalmerlowe committed:
fix: adds clear and delete item tests

Adds tests to confirm the proper functionality of the `clear()` method and the `__delitem__()` method. Increases robustness and ensures code coverage.

chalmerlowe committed:
chore: updates linting (removes white space)

chalmerlowe committed:
Merge branch 'main' into drop-cachetools-dep

chalmerlowe committed:
Merge branch 'main' into drop-cachetools-dep

chalmerlowe reviewed (approved):
LGTM

chalmerlowe commented on google/auth/_cache.py:
```diff
@@ -0,0 +1,64 @@
+from collections import OrderedDict
+
+
+class LRUCache(dict):
```
Daniel: 

Thanks for asking the question. The short answer is that `functools.lru_cache` is not the right tool for this job, despite a similarity in names.

**Incompatible Introspection**: The consuming code that uses `LRUCache` (in `google.auth.jwt.OnDemandCredentials`) uses the cache as a data store where it retrieves a token and then checks the expiration to confirm that the `expiry` < `helpers.utcnow()`. `functools.lru_cache` does not support retrieving a value without "using" it in a way that affects eviction, nor does it support inspecting the value (`expiry`) to decide if it is valid.

**Explicit Invalidation**: If the token is expired, the code needs to explicitly replace it. `functools.lru_cache` does not expose a way to invalidate or update a specific key; you can only clear the entire cache.

**Usage Pattern**: `functools.lru_cache` is a function decorator intended to hold the arguments for a decorated function so that the function does not need to be called again. Here, an object-based cache is needed to store state associated with the `OnDemandCredentials` instance.

chalmerlowe committed:
Merge branch 'main' into drop-cachetools-dep

chalmerlowe committed:
Merge branch 'main' into drop-cachetools-dep
